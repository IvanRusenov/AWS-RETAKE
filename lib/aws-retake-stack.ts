import * as cdk from 'aws-cdk-lib';
import {Construct} from 'constructs';
import {Bucket} from "aws-cdk-lib/aws-s3";
import {AttributeType, BillingMode, Table} from "aws-cdk-lib/aws-dynamodb";
import {Subscription, SubscriptionProtocol, Topic} from "aws-cdk-lib/aws-sns";
import {NodejsFunction} from "aws-cdk-lib/aws-lambda-nodejs";
import {Runtime} from "aws-cdk-lib/aws-lambda";
import * as path from "node:path";
import {Cors, LambdaIntegration, RestApi} from "aws-cdk-lib/aws-apigateway";
import {BucketDeployment, Source} from "aws-cdk-lib/aws-s3-deployment";
import {aws_iam, CfnOutput} from "aws-cdk-lib";

export class AwsRetakeStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);



        const adminRole = new aws_iam.Role(this, 'adminRole', {
            assumedBy: new aws_iam.CompositePrincipal(
                new aws_iam.ServicePrincipal('lambda.amazonaws.com'),
                new aws_iam.ServicePrincipal('apigateway.amazonaws.com'),
            ),
            managedPolicies: [
                aws_iam.ManagedPolicy.fromAwsManagedPolicyName('AdministratorAccess')
            ]
        });

// S3 Bucket for website
        const websiteBucket = new Bucket(this, 'CatLoversWebsiteBucket', {
            websiteIndexDocument: 'index.html',
            publicReadAccess: true,
            blockPublicAccess: {
                blockPublicAcls: false,
                blockPublicPolicy: false,
                ignorePublicAcls: false,
                restrictPublicBuckets: false
            },
            removalPolicy: cdk.RemovalPolicy.DESTROY,
        });

        const bucketDeployment = new BucketDeployment(this, "indexDeployment", {
            sources: [Source.asset(path.join(__dirname, "../resources"))],
            destinationBucket: websiteBucket
        });

        // DynamoDB Table (Single Table Design)
        const table = new Table(this, 'FavoriteCatsTable', {
            partitionKey: {name: 'PK', type: AttributeType.STRING},
            sortKey: {name: 'SK', type: AttributeType.STRING},
            billingMode: BillingMode.PAY_PER_REQUEST,
        });

        // SNS Topic for notifications
        const topic = new Topic(this, 'NewFavoriteCatTopic');

        new Subscription(this, "subscription", {
            topic: topic,
            protocol: SubscriptionProtocol.EMAIL,
            endpoint: "i_rusenov@abv.bg"
        });

        //Function
        const saveCatLambda = new NodejsFunction(this, "saveCatLambda", {
            handler: "handler",
            runtime: Runtime.NODEJS_20_X,
            entry: path.join(__dirname, "../src/saveCatLambda.ts"),
            role: adminRole,
            environment: {
                TABLE_NAME: table.tableName,
                TOPIC_ARN: topic.topicArn,
            }
        });

        // Grant permissions
        table.grantReadWriteData(saveCatLambda);
        topic.grantPublish(saveCatLambda);

        // API Gateway
        const api = new RestApi(this, 'CatLoversApi', {
            defaultCorsPreflightOptions: {
                allowOrigins: Cors.ALL_ORIGINS,
                allowMethods: Cors.ALL_METHODS,
                allowHeaders: ['Content-Type'],
                statusCode: 200
            },
            policy: new aws_iam.PolicyDocument({
                statements: [
                    new aws_iam.PolicyStatement({
                        actions: ['execute-api:Invoke'],
                        resources: ['execute-api:/*/*/*'],
                        principals: [new aws_iam.AnyPrincipal()],
                        effect: aws_iam.Effect.ALLOW,
                        conditions: {
                            'IpAddress': {
                                'aws:SourceIp': ['0.0.0.0/0']
                            }
                        }
                    })
                ]
            })
        });

        api.root.addResource("saveCat")
            .addMethod("POST", new LambdaIntegration(saveCatLambda, {proxy: true}));


        new CfnOutput(this, "websiteUrl", {
            key: "websiteUrl",
            value: websiteBucket.bucketWebsiteUrl
        });

        new CfnOutput(this, "ApiGatewayUrl", {
            key: "ApiGatewayUrl",
            value: api.url
        });

    }
}
