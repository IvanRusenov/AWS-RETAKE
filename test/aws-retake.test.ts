import * as cdk from 'aws-cdk-lib';
import {Template} from 'aws-cdk-lib/assertions';
import * as AwsRetake from '../lib/aws-retake-stack';

test('Snapshot test', () => {
    const app = new cdk.App();
    const stack = new AwsRetake.AwsRetakeStack(app, 'MyTestStack');
    const template = Template.fromStack(stack);

    expect(template).toMatchSnapshot();

});
