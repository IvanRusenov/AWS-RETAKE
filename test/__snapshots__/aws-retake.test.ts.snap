// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test 1`] = `
{
  "Outputs": {
    "ApiGatewayUrl": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "CatLoversApi6BBA0DE2",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "CatLoversApiDeploymentStageprod1FD8E93B",
            },
            "/",
          ],
        ],
      },
    },
    "CatLoversApiEndpoint33B53142": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "CatLoversApi6BBA0DE2",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "CatLoversApiDeploymentStageprod1FD8E93B",
            },
            "/",
          ],
        ],
      },
    },
    "websiteUrl": {
      "Value": {
        "Fn::GetAtt": [
          "CatLoversWebsiteBucketE8A5384A",
          "WebsiteURL",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CatLoversApi6BBA0DE2": {
      "Properties": {
        "Name": "CatLoversApi",
        "Policy": {
          "Statement": [
            {
              "Action": "execute-api:Invoke",
              "Condition": {
                "IpAddress": {
                  "aws:SourceIp": [
                    "0.0.0.0/0",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": {
                "AWS": "*",
              },
              "Resource": "execute-api:/*/*/*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "CatLoversApiAccountBA2D4930": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "CatLoversApi6BBA0DE2",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "CatLoversApiCloudWatchRole3149948E",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "CatLoversApiCloudWatchRole3149948E": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "CatLoversApiDeployment352B2FD741d533f0a5385c617243e8e8137f3458": {
      "DependsOn": [
        "CatLoversApiOPTIONS60AD6AF2",
        "CatLoversApisaveCatOPTIONSF2F3DAC2",
        "CatLoversApisaveCatPOST5BC5586C",
        "CatLoversApisaveCat26CDE60C",
      ],
      "Metadata": {
        "aws:cdk:do-not-refactor": true,
      },
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "CatLoversApi6BBA0DE2",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "CatLoversApiDeploymentStageprod1FD8E93B": {
      "DependsOn": [
        "CatLoversApiAccountBA2D4930",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "CatLoversApiDeployment352B2FD741d533f0a5385c617243e8e8137f3458",
        },
        "RestApiId": {
          "Ref": "CatLoversApi6BBA0DE2",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "CatLoversApiOPTIONS60AD6AF2": {
      "Properties": {
        "ApiKeyRequired": false,
        "AuthorizationType": "NONE",
        "HttpMethod": "OPTIONS",
        "Integration": {
          "IntegrationResponses": [
            {
              "ResponseParameters": {
                "method.response.header.Access-Control-Allow-Headers": "'Content-Type'",
                "method.response.header.Access-Control-Allow-Methods": "'OPTIONS,GET,PUT,POST,DELETE,PATCH,HEAD'",
                "method.response.header.Access-Control-Allow-Origin": "'*'",
              },
              "StatusCode": "200",
            },
          ],
          "RequestTemplates": {
            "application/json": "{ statusCode: 200 }",
          },
          "Type": "MOCK",
        },
        "MethodResponses": [
          {
            "ResponseParameters": {
              "method.response.header.Access-Control-Allow-Headers": true,
              "method.response.header.Access-Control-Allow-Methods": true,
              "method.response.header.Access-Control-Allow-Origin": true,
            },
            "StatusCode": "200",
          },
        ],
        "ResourceId": {
          "Fn::GetAtt": [
            "CatLoversApi6BBA0DE2",
            "RootResourceId",
          ],
        },
        "RestApiId": {
          "Ref": "CatLoversApi6BBA0DE2",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "CatLoversApisaveCat26CDE60C": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "CatLoversApi6BBA0DE2",
            "RootResourceId",
          ],
        },
        "PathPart": "saveCat",
        "RestApiId": {
          "Ref": "CatLoversApi6BBA0DE2",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "CatLoversApisaveCatOPTIONSF2F3DAC2": {
      "Properties": {
        "ApiKeyRequired": false,
        "AuthorizationType": "NONE",
        "HttpMethod": "OPTIONS",
        "Integration": {
          "IntegrationResponses": [
            {
              "ResponseParameters": {
                "method.response.header.Access-Control-Allow-Headers": "'Content-Type'",
                "method.response.header.Access-Control-Allow-Methods": "'OPTIONS,GET,PUT,POST,DELETE,PATCH,HEAD'",
                "method.response.header.Access-Control-Allow-Origin": "'*'",
              },
              "StatusCode": "200",
            },
          ],
          "RequestTemplates": {
            "application/json": "{ statusCode: 200 }",
          },
          "Type": "MOCK",
        },
        "MethodResponses": [
          {
            "ResponseParameters": {
              "method.response.header.Access-Control-Allow-Headers": true,
              "method.response.header.Access-Control-Allow-Methods": true,
              "method.response.header.Access-Control-Allow-Origin": true,
            },
            "StatusCode": "200",
          },
        ],
        "ResourceId": {
          "Ref": "CatLoversApisaveCat26CDE60C",
        },
        "RestApiId": {
          "Ref": "CatLoversApi6BBA0DE2",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "CatLoversApisaveCatPOST5BC5586C": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "saveCatLambda1F4F7FEF",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "CatLoversApisaveCat26CDE60C",
        },
        "RestApiId": {
          "Ref": "CatLoversApi6BBA0DE2",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "CatLoversApisaveCatPOSTApiPermissionMyTestStackCatLoversApi490886BEPOSTsaveCat3620B76E": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "saveCatLambda1F4F7FEF",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "CatLoversApi6BBA0DE2",
              },
              "/",
              {
                "Ref": "CatLoversApiDeploymentStageprod1FD8E93B",
              },
              "/POST/saveCat",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "CatLoversApisaveCatPOSTApiPermissionTestMyTestStackCatLoversApi490886BEPOSTsaveCat3CF23E60": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "saveCatLambda1F4F7FEF",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "CatLoversApi6BBA0DE2",
              },
              "/test-invoke-stage/POST/saveCat",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "CatLoversWebsiteBucketE8A5384A": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": false,
          "BlockPublicPolicy": false,
          "IgnorePublicAcls": false,
          "RestrictPublicBuckets": false,
        },
        "Tags": [
          {
            "Key": "Administrator",
            "Value": "Ivan",
          },
          {
            "Key": "aws-cdk:cr-owned:b9b90c61",
            "Value": "true",
          },
        ],
        "WebsiteConfiguration": {
          "IndexDocument": "index.html",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "CatLoversWebsiteBucketPolicyE930AF87": {
      "Properties": {
        "Bucket": {
          "Ref": "CatLoversWebsiteBucketE8A5384A",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "CatLoversWebsiteBucketE8A5384A",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": {
      "DependsOn": [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "9a1fcb4a7ecba81ad70e9d3fb241f6794497da945dae5f25924e4dd002b65f2d.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_CA_BUNDLE": "/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem",
          },
        },
        "Handler": "index.handler",
        "Layers": [
          {
            "Ref": "indexDeploymentAwsCliLayer7772B932",
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.11",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "CatLoversWebsiteBucketE8A5384A",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "CatLoversWebsiteBucketE8A5384A",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": [
          {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "FavoriteCatsTable64C4E88A": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "PK",
            "AttributeType": "S",
          },
          {
            "AttributeName": "SK",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "PK",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "SK",
            "KeyType": "RANGE",
          },
        ],
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "NewFavoriteCatTopicC6940A17": {
      "Properties": {
        "Tags": [
          {
            "Key": "Administrator",
            "Value": "Ivan",
          },
        ],
      },
      "Type": "AWS::SNS::Topic",
    },
    "adminRoleC345D70B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AdministratorAccess",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "adminRoleDefaultPolicy4FE403EF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "FavoriteCatsTable64C4E88A",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "NewFavoriteCatTopicC6940A17",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "adminRoleDefaultPolicy4FE403EF",
        "Roles": [
          {
            "Ref": "adminRoleC345D70B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "indexDeploymentAwsCliLayer7772B932": {
      "Properties": {
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "1b2c92f2cd21c170884393633bd4c732676df6290562199b6e3ca5e2a1be7d18.zip",
        },
        "Description": "/opt/awscli/aws",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "indexDeploymentCustomResource5A9A1D16": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DestinationBucketName": {
          "Ref": "CatLoversWebsiteBucketE8A5384A",
        },
        "OutputObjectKeys": true,
        "Prune": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": [
          {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
        ],
        "SourceObjectKeys": [
          "fcc31f6294517cb72cbd6f700878760194fff5282180da32c56b2d785b8d2271.zip",
        ],
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
    "saveCatLambda1F4F7FEF": {
      "DependsOn": [
        "adminRoleDefaultPolicy4FE403EF",
        "adminRoleC345D70B",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "8bfdcc3ecf9968437e5bc45409c41b7d8a3cade638cc0a60875c0e84ada79a14.zip",
        },
        "Environment": {
          "Variables": {
            "TABLE_NAME": {
              "Ref": "FavoriteCatsTable64C4E88A",
            },
            "TOPIC_ARN": {
              "Ref": "NewFavoriteCatTopicC6940A17",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "adminRoleC345D70B",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "subscriptionD5B264BF": {
      "Properties": {
        "Endpoint": "i_rusenov@abv.bg",
        "Protocol": "email",
        "TopicArn": {
          "Ref": "NewFavoriteCatTopicC6940A17",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
